muxers/mplex/src/io.rs

```diff
@@ -0,0 +1,733 @@
+                let id = frame.local_id();
+                if self.can_read(&id) || frame.is_open() {
+                    trace!("Buffering frame: {:?} (total: {})", frame, self.buffer.len() + 1);
+                    self.buffer.push(frame);
+                    self.notifier_read.wake_by_id(id);
+                } else if frame.is_data() {
+                    debug!("Dropping {:?} for closed or unknown substream: {:?}", frame, id);
+                }
+            }
+        }
+    }
+
+    /// Generates the next outbound stream ID.
+    fn next_outbound_stream_id(&mut self) -> LocalStreamId {
+        let id = self.next_outbound_stream_id;
+        self.next_outbound_stream_id = self.next_outbound_stream_id.next();
+        id
+    }
+
+    /// Checks whether a substream is open for reading.
+    fn can_read(&self, id: &LocalStreamId) -> bool {
+        match self.open_substreams.get(id) {
+            Some(SubstreamState::Open) | Some(SubstreamState::SendClosed) => true,
+            _ => false,
+        }
+    }
+
+    /// Sends pending frames, without flushing.
+    fn send_pending_frames(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        while let Some(frame) = self.pending_frames.pop() {
+            if self.poll_send_frame(cx, || frame.clone())?.is_pending() {
+                self.pending_frames.push(frame);
+                return Poll::Pending
+            }
+        }
+
+        Poll::Ready(Ok(()))
+    }
+
+    /// Records a fatal error for the multiplexed I/O stream.
+    fn on_error<T>(&mut self, e: io::Error) -> Poll<io::Result<T>> {
+        log::debug!("Multiplexed connection failed: {:?}", e);
+        self.status = Status::Err(io::Error::new(e.kind(), e.to_string()));
+        self.pending_frames =  Vec::new();
+        self.open_substreams = Default::default();
+        self.buffer = Default::default();
+        Poll::Ready(Err(e))
+    }
+
+    /// Checks that the multiplexed stream has status `Ok`,
+    /// i.e. is not closed and did not encounter a fatal error.
+    fn guard_ok(&self) -> io::Result<()> {
+        match &self.status {
+            Status::Closed => Err(io::Error::new(io::ErrorKind::Other, "Connection is closed")),
+            Status::Err(e) => Err(io::Error::new(e.kind(), e.to_string())),
+            Status::Ok => Ok(())
+        }
+    }
+}
+
+/// The operating states of a substream.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+enum SubstreamState {
+    /// An `Open` frame has been received or sent.
+    Open,
+    /// A `Close` frame has been sent, but the stream is still open
+    /// for reading (half-close).
+    SendClosed,
+    /// A `Close` frame has been received but the stream is still
+    /// open for writing (remote half-close).
+    RecvClosed
+}
+
+struct NotifierRead {
+    /// List of wakers to wake when read operations can proceed
+    /// on a substream (or in general, for the key `None`).
+    pending: Mutex<FnvHashMap<Option<LocalStreamId>, Waker>>,
```

Comment:
If I understand correctly, wakers whose key (the `Option<LocalStreamId>`) is `Some` come from calling `poll_read_stream`, and wakers whose key is `None` come from calling `poll_next_stream`.

If that's accurate, then it is correct to have a `Waker` here rather than a `Vec<Waker>`.

I find it quite hard to follow the logic, and it doesn't seem very fool-proof, but I also don't know how to make it clearer.

Reply:
> If that's accurate, then it is correct to have a Waker here rather than a Vec<Waker>.

Yes.

> I find it quite hard to follow the logic, and it doesn't seem very fool-proof, but I also don't know how to make it clearer.

Not sure what logic you mean, I guess in general the task wakeup logic for reading? Yes, that is, as is mostly the case, subtle. There are these two different "threads of control" for reading by design: 1) Awaiting the next inbound stream and b) Awaiting new data on a specific stream. Thereby 1) and each instance of 2) potentially read and buffer frames for other streams (or for new inbound streams) that need to result in wakeup attempts.

Reply:
> Can one of you expand on why it is safe to only cache a single waker for Option<LocalStreamId>::None?

Because the key `None` is only used by `poll_next_stream`, which is part of the implementation for `StreamMuxer::poll_event` whose API contract explicitly states that "Only the latest task that was used to call this method may be notified.". The API contract for `StreamMuxer::read_substream`, `StreamMuxer::write_substream` etc. is similar. In general, and as far as I know, the reason for `StreamMuxer`s to be `Sync` is to permit using different substreams from different threads, not to use the same substream from different threads. Only waking the last task that `poll`s is generally common for most such APIs, e.g. [Future::poll](https://docs.rs/futures/0.3.5/futures/future/trait.Future.html#tymethod.poll) has an analogous contract. In practice, I would think the reason for this contract is a matter of practicality due ease and efficiency of implementation and the relatively rare usefulness for having to "remember" all tasks to wake, instead of just the last.

Reply:
Thanks for the in-depth explanation.

Reply:
> I would think the reason for this contract is a matter of practicality due to ease and efficiency of implementation and the relatively rare usefulness for having to "remember" all tasks to wake, instead of just the last.

One big problem with remember all tasks to wake up is that your container of wakers might grow a lot.

For instance, imagine that it takes 60 seconds before data is received on a substream and that, due to the nature of polling, we poll a substream 10 times per second. This results in 600 calls to `poll_read_substream` before the wakers actually get awakened. This means that we would insert 600 elements in the list of wakers.

The [`will_wake`](https://doc.rust-lang.org/nightly/std/task/struct.Waker.html#method.will_wake) method allows reducing the amount of duplicates, and also we're not really supposed to poll a substream 10 times per second. But both these mechanisms are unreliable.

Enforcing one waker per "thing to wake" guarantees a bound to the number of wakers, which is not a bad idea.